### ISSUE TITLE
Read and Write property error

### ISSUE THREAD

**Issue by Nicolas3347** [jci-metasys/basic-services-dotnet#132]
Doing the same calls for an object works fine at the beginning and crashes after a while. I do verify that the access token is valid otherwise log in again.
I cannot explain why this is working perfectly and then crash after a while. Any idea?

Read Property
```
Base: JohnsonControls.Metasys.BasicServices.MetasysHttpException: Call failed with status code 500 (Internal Server Error): GET https://tdcbas001p/api/v3/objects/2374ae8e-b15e-523c-b0eb-834b84327822/attributes/presentValue
   at JohnsonControls.Metasys.BasicServices.BasicServiceProvider.ThrowHttpException(FlurlHttpException e)
   at JohnsonControls.Metasys.BasicServices.MetasysClient.ReadPropertyAsync(Guid id, String attributeName)
```

Writer Property
```
Base: JohnsonControls.Metasys.BasicServices.MetasysHttpException: Call failed with status code 500 (Internal Server Error): PATCH https://tdcbas001p/api/v3/objects/2374ae8e-b15e-523c-b0eb-834b84327822
   at JohnsonControls.Metasys.BasicServices.BasicServiceProvider.ThrowHttpException(FlurlHttpException e)
   at JohnsonControls.Metasys.BasicServices.MetasysClient.WritePropertyRequestAsync(Guid id, Dictionary`2 body)
   at JohnsonControls.Metasys.BasicServices.MetasysClient.WritePropertyAsync(Guid id, String attributeName, Object newValue)
   at JohnsonControls.Metasys.BasicServices.MetasysClient.WriteProperty(Guid id, String attributeName, Object newValue)
```

**Comment by michaelgwelch** [comment]
@Nicolas3347 I recommend you contact your local field service support for this issue. This does not appear to be an issue with the `basic-services-dotnet` library but rather an issue on your server. 

How many calls are you making and how often?

**Comment by Nicolas3347** [comment]
I do make about 75 to 100 calls a day usually from 7pm to 8am, so night stuff.
The access token is valid for about 9 hours as from the AccessToken.Expires date, but I do a client.RefreshAsync() if I hit the Expiration Date to renew the token.
My Login code:

```
private async Task<bool> Login(string DomainNameOrIP, string UserName, string Password)
    {
        try
        {
            //return true;// for testing purpose
            //// Automatically refresh token using plain credentials

            // ***** START METASYS ****** \\
            ////// Automatically refresh token using plain credentials
            await client.TryLoginAsync(username: UserName, password: Password).ConfigureAwait(false);
            
            ////// Do not automatically refresh token using plain credentials
            ////client.TryLogin(UserName, Password, false);
            
            accessToken = client.GetAccessToken();
            // *****  END METASYS ****** \\

            errorHandler(Message: "Getting Token: " + accessToken?.Token + " --- Expires: " + accessToken?.Expires, FunctionName: this.GetType().ToString() + ".Login", eventId: eventId.AddInReturnValue, eventLogEntry: EventLogEntryType.Warning);

            //Console.WriteLine(accessToken);
            /////*        
            ////{
            ////    "Issuer": "metasysserver",
            ////    "IssuedTo": "metasysapiuser",
            ////    "Token": "Bearer eyJ0eXAiOiJKV1QiLCJhbGciOiJSUzI1NiIsIng1dCI6IkJzR0lWelVZcjN0MkI0RGRtT1ljMTdBLVZJOCIsImtpZCI6IkJzR0lWelVZcjN0MkI0RGRtT1ljMTdBLVZJOCJ9.eyJpc3MiOiJodHRwOi8vbG9jYWxob3N0Ojk1MDYvQVBJLkF1dGhlbnRpY2F0aW9uU2VydmljZSIsImF1ZCI6Imh0dHA6Ly9sb2NhbGhvc3Q6OTUwNi9BUEkuQXV0aGVudGljYXRpb25TZXJ2aWNlL3Jlc291cmNlcyIsImV4cCI6MTU4OTI5MzEzMCwibmJmIjoxNTg5MjkxMzMwLCJjbGllbnRfaWQiOiJtZXRhc3lzX3VpIiwic2NvcGUiOlsibWV0YXN5c19hcGkiLCJvZmZsaW5lX2FjY2VzcyIsIm9wZW5pZCJdLCJzdWIiOiI4ZGE4YjE4Yy1lMTk1LTRlMmMtOGU2Zi0zNTE2Zjc0ZWFhNGIiLCJhdXRoX3RpbWUiOjE1ODkyOTEzMzAsImlkcCI6Imlkc3J2IiwiVXNlcklkIjoiMSIsIlVzZXJOYW1lIjoibWV0YXN5c3N5c2FnZW50IiwiSXNBZG1pbiI6IlRydWUiLCJJc1Bhc3N3b3JkQ2hhbmdlUmVxdWlyZWQiOiJGYWxzZSIsIklzVGVybXNBbmRDb25kaXRpb25zUmVxdWlyZWQiOiJGYWxzZSIsIkxpY2Vuc2VJbmZvIjoie1wiSXNMaWNlbnNlZFwiOnRydWUsXCJMaWNlbnNlVHlwZVwiOlwiZ3JhY2VcIn0iLCJDdWx0dXJlIjoiZW4tVVMiLCJhbXIiOlsicGFzc3dvcmQiXX0.egzw1bs1831pEBWWXbBOYWGU5wFsI3sEnL7RgCIHHbHmcxtpPPqLq54znpoUoLFrMUeymZj5rkrt_mF-CNIpCE3halZNAH-CR1U46LTZi5CMaDfYlP-wHxikAGV5GwFjlHjGNOUaFtd7n4yC5sH08pHQfXXD5gKDm_FVMfUJXAo-E8gmrkU0wMn5U2FRyQyj7Yhq6jaj7MPTF__Xz46sG3WtDr45WK2NmuwiLDv408URZ5fJxlMngRpjSIONHVAIwna_H0AguHiIELkvuRVYcRqIH5kb1YdFt-3fsnTV9xwpozZZ44dh-4I7x466I-UGlLHAnScWILUbPcpRNWm0Uw",
            ////    "Expires": "2020-05-12T14:18:51Z"
            ////}
            ////*/

            AddinLoaded = accessToken!=null;
            return AddinLoaded;
        }
        catch (Exception ex)
        {
            errorHandler(ex, this.GetType().ToString() + ".Login");
            AddinLoaded = false;
            return false;
        }
    }
```    

and my verification on each calls:

```
private async Task isTokenExpired()
    {
        // ***** START METASYS ****** \\

        //"Expires": "2020-05-12T14:18:51Z"
        //Check if accessToken isn't null or is expired now otherwise Refresh the client again to create a new token
        if (accessToken == null || accessToken.Expires < DateTime.Now)
        {
            accessToken = await client.RefreshAsync();
            errorHandler(Message: "After Refreshing Token - Token: " + accessToken?.Token + " --- Expires: " + accessToken?.Expires, FunctionName: this.GetType().ToString() + ".isTokenExpired", eventId: eventId.AddInReturnValue, eventLogEntry: EventLogEntryType.Warning);
            
        }

        //Check if accessToken is ok now otherwise login again to create a new token
        if (accessToken == null)
        {
            isLogin = await Login(hostname, USR, PWD);
            AddinLoaded = isLogin;
        }
        // *****  END METASYS ****** \\
    }
```


"but rather an issue on your server." you mean the Metasys server. My App is running on a different server connecting via our network.
The problem is that this is working fine when I start my app and a few calls go through but then doesn't work after a few minutes or hours, and never starts again if not restarting my app.
My suspicion is that must be something around the connection or token timing out.

Thanks for helping.

**Comment by Nicolas3347** [comment]
Thanks for your input with various errors code. I did encounter the 401 (Authorized) which drove me to that Access Token and resolve it by refreshAsync() before calls I think for now.
As for the error 500, it is issued by Metasys Dll as per the error message, isn't it?
```
Base: JohnsonControls.Metasys.BasicServices.MetasysHttpException: Call failed with status code 500 (Internal Server Error): GET https://tdcbas001p/api/v3/objects/2374ae8e-b15e-523c-b0eb-834b84327822/attributes/presentValue
   at JohnsonControls.Metasys.BasicServices.BasicServiceProvider.ThrowHttpException(FlurlHttpException e)
   at JohnsonControls.Metasys.BasicServices.MetasysClient.ReadPropertyAsync(Guid id, String attributeName)
```
And the Error is generated by this BasicServices.MetasysClient call reaching the Metasys server
```
public async Task<Variant> ReadPropertyAsync(Guid id, string attributeName)
        {
            JToken response = null;
            Variant result = new Variant();
            try
            {
                response = await Client.Request(new Url("objects").AppendPathSegments(id, "attributes", attributeName)).GetJsonAsync<JToken>().ConfigureAwait(continueOnCapturedContext: false);
                result = new Variant(id, response, attributeName, Culture, Version);
                return result;
            }
            catch (FlurlHttpException e)
            {
                ThrowHttpException(e);
                return result;
            }
            catch (NullReferenceException inner)
            {
                throw new MetasysPropertyException(response.ToString(), inner);
            }
        }
```

I did try to reach the local person in charge of the Metasys server, but no response, plus I don't know if they know enough about programming :-) 

Again Thank you for your help.

**Comment by michaelgwelch** [comment]
> As for the error 500, it is issued by Metasys Dll as per the error message, isn't it?

No, that's the http status code returned by the Metasys Server. The dll in this repo makes http calls to the server and the server returns standard http status codes.

**Comment by michaelgwelch** [comment]
> I did try to reach the local person in charge of the Metasys server, but no response, plus I don't know if they know enough about programming :-)

They should be able to open a ticket with Metasys Field Support, or they should know someone who knows how to contact our field support.

**Comment by Nicolas3347** [comment]
Hi Michael,
Did reach local support and now answer yet.
However, would it be correct to assumed that, there is a disconnect between the Metasys IIS server and the Access Token.

If I do client.Login() at 13:00:00
I got an Access Token expiring at ~ 22:00:00
However the IIS Webserver Session might be expired after ~30 minutes from the last activity call. 
So event when I do a client.refresh() about few hours later (may be 3) it does crash with internal error return.

If my assumption is good and make sense we should setup a way longer IIS Session expiring date/time.

The way I kind of fixed it, is that I request the servertime function and if I got an error then I reconnect and re-login and then call my function (writeproperty) again.
